packages/shared/src/authz/policies/project/canInvite.ts

```diff
@@ -0,0 +1,58 @@
+import { err, ok } from 'true-myth/result'
+import { MaybeUserContext, ProjectContext } from '../../domain/context.js'
+import { AuthCheckContextLoaderKeys } from '../../domain/loaders.js'
+import { AuthPolicy } from '../../domain/policies.js'
+import {
+  ServerNoAccessError,
+  ServerNoSessionError,
+  WorkspaceSsoSessionNoAccessError,
+  WorkspaceNoAccessError,
+  ProjectNoAccessError,
+  ProjectNotFoundError,
+  ServerNotEnoughPermissionsError,
+  ProjectNotEnoughPermissionsError
+} from '../../domain/authErrors.js'
+import { ensureMinimumServerRoleFragment } from '../../fragments/server.js'
+import { Roles } from '../../../core/constants.js'
+import { ensureMinimumProjectRoleFragment } from '../../fragments/projects.js'
+
+type PolicyLoaderKeys =
+  | typeof AuthCheckContextLoaderKeys.getEnv
+  | typeof AuthCheckContextLoaderKeys.getServerRole
+  | typeof AuthCheckContextLoaderKeys.getProject
+  | typeof AuthCheckContextLoaderKeys.getProjectRole
+  | typeof AuthCheckContextLoaderKeys.getWorkspace
+  | typeof AuthCheckContextLoaderKeys.getWorkspaceRole
+  | typeof AuthCheckContextLoaderKeys.getWorkspaceSsoProvider
+  | typeof AuthCheckContextLoaderKeys.getWorkspaceSsoSession
+
+type PolicyArgs = MaybeUserContext & ProjectContext
+
+type PolicyErrors =
+  | InstanceType<typeof ServerNoAccessError>
+  | InstanceType<typeof ServerNoSessionError>
+  | InstanceType<typeof ServerNotEnoughPermissionsError>
+  | InstanceType<typeof ProjectNoAccessError>
+  | InstanceType<typeof ProjectNotFoundError>
+  | InstanceType<typeof ProjectNotEnoughPermissionsError>
+  | InstanceType<typeof WorkspaceSsoSessionNoAccessError>
+  | InstanceType<typeof WorkspaceNoAccessError>
+
+export const canInvitePolicy: AuthPolicy<PolicyLoaderKeys, PolicyArgs, PolicyErrors> =
+  (loaders) =>
+  async ({ userId, projectId }) => {
+    const ensuredServerRole = await ensureMinimumServerRoleFragment(loaders)({
+      userId,
+      role: Roles.Server.User
+    })
+    if (ensuredServerRole.isErr) return err(ensuredServerRole.error)
+
+    const ensuredProjectRole = await ensureMinimumProjectRoleFragment(loaders)({
```

Comment:
Is this really enough? Don't we also need to check that the user has access to the workspace that the project is in + has an SSO session in it?

I think this entire thing can be replaced w/ `ensureImplicitProjectMemberWithWriteAccessFragment()` w/ the Owner role passed in through args

Reply:
Fair, updated to use fragment.